src/main/java/mate/academy/springboot/datajpa/controller/ProductController.java

```diff
@@ -0,0 +1,76 @@
+package mate.academy.springboot.datajpa.controller;
+
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import mate.academy.springboot.datajpa.dto.request.ProductRequestDto;
+import mate.academy.springboot.datajpa.dto.response.ProductResponseDto;
+import mate.academy.springboot.datajpa.model.Product;
+import mate.academy.springboot.datajpa.service.ProductService;
+import mate.academy.springboot.datajpa.service.mapper.ProductMapper;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/products")
+public class ProductController {
+    private final ProductMapper productMapper;
+    private final ProductService productService;
+
+    public ProductController(ProductMapper productMapper, ProductService productService) {
+        this.productMapper = productMapper;
+        this.productService = productService;
+    }
+
+    @PostMapping
+    public ProductResponseDto create(@RequestBody ProductRequestDto productRequestDto) {
+        Product product = productService.add(productMapper.toModel(productRequestDto));
+        return productMapper.toDto(product);
+    }
+
+    @PutMapping("/{id}")
+    public ProductResponseDto update(@PathVariable Long id,
+                                     @RequestBody ProductRequestDto productRequestDto) {
+        Product product = productMapper.toModel(productRequestDto);
+        product.setId(id);
+        Product save = productService.add(product);
+        return productMapper.toDto(save);
+    }
+
+    @GetMapping("/{id}")
+    public ProductResponseDto get(@PathVariable Long id) {
+        Product product = productService.getById(id);
+        return productMapper.toDto(product);
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        productService.deleteById(id);
+    }
+
+    @GetMapping
+    public List<ProductResponseDto> getAllByCategories(@RequestParam List<String> categories) {
+        return categories.stream()
+                .map(productService::findAllByCategories)
+                .flatMap(Collection::stream)

```

Comment:
you should rewrite your repository method. We have the possibility to pass several categories to JPA repository query. So you can get results in one request to the database.